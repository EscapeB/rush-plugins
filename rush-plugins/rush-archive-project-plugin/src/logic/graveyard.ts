import * as path from 'path';
import { PackageName, IParsedPackageName, JsonFile, FileSystem } from '@rushstack/node-core-library';
import { RushConfiguration } from '@rushstack/rush-sdk';
import { loadRushConfiguration } from './rushConfiguration';

export const defaultGraveyardRelativeFolder: string = 'common/_graveyard';

export interface IGraveyardInfo {
  graveyardRelativeFolder: string;
  tarballFolder: string;
  tarballRelativeFolder: string;
  tarballName: string;
}

export interface IGetGraveyardInfoParams {
  monoRoot: string;
  packageName: string;
}

export interface IArchiveConfig {
  graveyardFolder?: string;
}

export const getGraveyardInfo = ({ monoRoot, packageName }: IGetGraveyardInfoParams): IGraveyardInfo => {
  const rushConfiguration: RushConfiguration = loadRushConfiguration();
  // Check for config file
  const pluginOptionsJsonFilePath: string = path.join(
    rushConfiguration.rushPluginOptionsFolder,
    `rush-archive-project-plugin.json`
  );

  let graveyardRelativeFolder: string = defaultGraveyardRelativeFolder;
  let archiveConfigs: IArchiveConfig | undefined;
  try {
    archiveConfigs = JsonFile.load(pluginOptionsJsonFilePath);
  } catch (e) {
    if (!FileSystem.isNotExistError(e as Error)) {
      throw e;
    }
  }

  if (archiveConfigs?.graveyardFolder) {
    graveyardRelativeFolder = archiveConfigs.graveyardFolder;
  }

  const parsedPackageName: IParsedPackageName = PackageName.parse(packageName);
  const tarballRelativeFolder: string = path.join(graveyardRelativeFolder, parsedPackageName.scope);
  const tarballFolder: string = path.join(monoRoot, tarballRelativeFolder);
  return {
    graveyardRelativeFolder,
    tarballFolder,
    tarballRelativeFolder,
    tarballName: `${parsedPackageName.unscopedName}.tar.gz`
  };
};
