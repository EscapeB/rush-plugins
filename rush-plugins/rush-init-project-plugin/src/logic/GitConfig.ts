import { Executable } from "@rushstack/node-core-library";
import type { SpawnSyncReturns } from "child_process";

export const getGitUserName = (cwd: string): string => {
  try {
    const result: SpawnSyncReturns<string> =  Executable.spawnSync(
      'git',
      ['config', 'user.name'],
      {
        currentWorkingDirectory: cwd,
        stdio: "pipe",
      }
    );
    if (result.status !== 0) {
      return '';
    }
    return result.stdout?.trim() ?? '';
  } catch (e) {
    // no-catch
    return '';
  }
}