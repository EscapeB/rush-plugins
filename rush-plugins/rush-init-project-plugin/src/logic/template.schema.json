{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "Template config File",
  "description": "The main configuration file Template management.",
  "type": "object",
  "properties": {
    "$schema": {
      "description": "Part of the JSON Schema standard, this optional keyword declares the URL of the schema that the file conforms to. Editors may download the schema and use it to perform syntax highlighting.",
      "type": "string"
    },
    "templates": {
      "description": "A list of templates managed by this tool.",
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "displayName": {
            "description": "The display name of the template.",
            "type": "string"
          },
          "templateFolder": {
            "description": "The path to the template folder relative to the template config file.",
            "type": "string"
          }
        },
        "additionalProperties": false,
        "required": ["templateFolder"]
      }
    },
    "globalPluginEntryPoint": {
      "description": "entry point for global plugins",
      "type": "string"
    }
  },
  "additionalProperties": false,
  "required": ["templates"]
}
