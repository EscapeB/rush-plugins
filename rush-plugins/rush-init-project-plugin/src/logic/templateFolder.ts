import * as path from 'path';
import { FileSystem } from '@rushstack/node-core-library';
import { loadRushConfiguration } from './loadRushConfiguration';

export interface ITemplatePathNameType {
  templateFolder: string;
  displayName?: string;
}

const templatesFolder2templateNameList: Record<string, Array<ITemplatePathNameType>> = {};

export const getTemplatesFolder = (): string => {
  const { commonFolder } = loadRushConfiguration();
  const templatesFolder: string = path.join(commonFolder, '_templates');
  return templatesFolder;
};

export const getTemplatesFolderAndValidate = (): string => {
  const templatesFolder: string = getTemplatesFolder();
  if (!FileSystem.exists(templatesFolder)) {
    FileSystem.ensureFolder(templatesFolder);
    throw new Error(`Templates folder created, please setup template under "${templatesFolder}"`);
  }

  const templateNameList: ITemplatePathNameType[] = getTemplateNameList(templatesFolder);
  if (templateNameList.length === 0) {
    throw new Error(`Please setup template under ${templatesFolder}`);
  }

  return templatesFolder;
};

// get template list from folder
export function getTemplateNameList(templatesFolder: string): ITemplatePathNameType[] {
  let templateNameList: ITemplatePathNameType[] = templatesFolder2templateNameList[templatesFolder];
  if (!templateNameList) {
    const templateFolderList: string[] = FileSystem.readFolder(templatesFolder)
      .filter((filename: string) => {
        return FileSystem.getStatistics(path.resolve(templatesFolder, filename)).isDirectory();
      })
      .filter((filename: string) => {
        return !filename.startsWith('_');
      });
    templateNameList = templateFolderList.map((templateFolder: string) => ({ templateFolder }));
  }
  // eslint-disable-next-line require-atomic-updates
  templatesFolder2templateNameList[templatesFolder] = templateNameList;
  return templateNameList;
}

export function getTemplateFolder(template: string): string {
  const templatesFolder: string = getTemplatesFolder();
  const templateFolder: string = path.join(templatesFolder, template);
  return templateFolder;
}
