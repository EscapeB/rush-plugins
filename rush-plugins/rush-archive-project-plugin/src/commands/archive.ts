import type {
  RushConfiguration,
  RushConfigurationProject,
} from "@rushstack/rush-sdk";
import { FileSystem, JsonFile, JsonObject } from "@rushstack/node-core-library";
import * as path from "path";
import * as tar from "tar";
import { getCheckpointBranch, gitCheckIgnored, gitFullClean, pushGitBranch } from "../logic/git";
import { getGraveyardInfo } from "../logic/graveyard";
import { IProjectCheckpointMetadata, ProjectMetadata } from "../logic/projectMetadata";
import ora from "ora";
import inquirer from 'inquirer';
import fs from 'fs';
import { loadRushConfiguration } from "../logic/rushConfiguration";
import { convertProjectMetadataToMD } from "../logic/generateMD";

interface IArchiveConfig {
  packageName: string;
  gitCheckpoint: boolean;
  shouldPushGitCheckpoint?: boolean;
}

export async function archive({ packageName, gitCheckpoint, shouldPushGitCheckpoint }: IArchiveConfig): Promise<void> {
  let spinner: ora.Ora | undefined;
  const rushConfiguration: RushConfiguration = loadRushConfiguration();
  const monoRoot: string = rushConfiguration.rushJsonFolder;
  const project: RushConfigurationProject | undefined =
    rushConfiguration.getProjectByName(packageName);
  if (!project) {
    throw new Error(`Could not find project with package name ${packageName}`);
  }

  // check project is depended by other projects
  const consumingProjectNames: string[] = Array.from(
    project.consumingProjects
  ).map((x: RushConfigurationProject) => x.packageName);
  if (consumingProjectNames.length) {
    throw new Error(`Target project ${packageName} is depended by other ${
      consumingProjectNames.length
    } project(s):
${consumingProjectNames.join(", ")}`);
  }

  const { projectFolder, projectRelativeFolder } = project;

  // git clean -xdf
  spinner = ora(`Cleaning ${projectRelativeFolder}`).start();
  gitFullClean(projectFolder);
  spinner.succeed(`Clean ${projectRelativeFolder} complete`);

  const { graveyardRelativeFolder, tarballRelativeFolder, tarballFolder, tarballName } =
    getGraveyardInfo({
      monoRoot,
      packageName,
    });
  FileSystem.ensureFolder(tarballFolder);

  if (gitCheckpoint) {
    spinner = ora('Attempting to create a git checkpoint branch');
    const branchName: string = getCheckpointBranch(rushConfiguration.rushJsonFolder,packageName);
    spinner.succeed(`Git Checkpoint created at branch: ${branchName}`);
    let shouldPushBranch: boolean;
    if (shouldPushGitCheckpoint === undefined) {
      // Prompt user to push the newly created branch
      const { pushBranch } = await inquirer.prompt([
        { type: 'confirm', name: 'pushBranch', message: 'Push checkpoint branch to remote?' }
      ])
      shouldPushBranch = pushBranch;
    } else {
      shouldPushBranch = shouldPushGitCheckpoint;
    }

    if (shouldPushBranch) {
      pushGitBranch(rushConfiguration.rushJsonFolder, branchName);
    }
    // Add data to metadata file
    const archivedProjectMetadataFilePath: string = path.join(monoRoot, graveyardRelativeFolder, 'projectCheckpoints.json');
    const archivedProjectMetadataMdFilePath: string = path.join(monoRoot, graveyardRelativeFolder, 'projectCheckpoints.md');
    let archivedProjectMetadataObject: { [key in string]: IProjectCheckpointMetadata } = {};
    if (FileSystem.exists(archivedProjectMetadataFilePath)) {
      archivedProjectMetadataObject = JsonFile.load(archivedProjectMetadataFilePath);
    }
    // Try to pull a description from the package's package.json file
    let description: string = "";
    const projectJson: string = `${project.projectFolder}/package.json`;
    if (FileSystem.exists(projectJson)) {
      const packageJson: any = JsonFile.load(`${project.projectFolder}/package.json`);
      description = packageJson.description || "";
    }
    archivedProjectMetadataObject[packageName] = {
      checkpointBranch: branchName,
      archivedOn: new Date().toISOString(),
      description
    }
    JsonFile.save(archivedProjectMetadataObject, archivedProjectMetadataFilePath);

    fs.writeFileSync(archivedProjectMetadataMdFilePath, convertProjectMetadataToMD(archivedProjectMetadataObject));
  }

  // create a metadata.json file
  spinner = ora(`Creating metadata.json for ${projectRelativeFolder}`).start();
  const rawRushJson: JsonObject = JsonFile.load(rushConfiguration.rushJsonFile);
  const rawProjectConfig: JsonObject = rawRushJson.projects.find(
    (x: JsonObject) => x.packageName === packageName
  );
  const projectMetadata: ProjectMetadata = new ProjectMetadata(
    rawProjectConfig
  );

  const projectMetadataFilepath: string = path.join(
    projectFolder,
    ProjectMetadata.FILENAME
  );
  projectMetadata.save(projectMetadataFilepath);
  spinner.succeed(`Created metadata.json for ${projectRelativeFolder}`);

  // create archive tarball
  spinner = ora(`Creating tarball for ${projectRelativeFolder}`).start();
  try {
    //tar -czf test.tar.gz -C project_relative_folder .
    tar.create(
      {
        gzip: true,
        file: tarballName,
        sync: true,
        cwd: projectFolder,
      },
      ["."]
    );
  } catch (e: any) {
    throw new Error(`Failed to create tarball: ${e.message}`);
  }
  spinner.succeed(`Created tarball for ${projectRelativeFolder}`);

  // move the tarball to the graveyard folder
  spinner = ora(`Moving tarball to ${tarballRelativeFolder}`).start();
  const finalTarballPath: string = path.join(tarballFolder, tarballName);
  FileSystem.move({
    sourcePath: tarballName,
    destinationPath: finalTarballPath,
  });
  spinner.succeed(`Moved tarball to ${tarballRelativeFolder}`);

  // check if the tarball is ignored by git
  spinner = ora(`Checking if tarball is ignored by git`).start();
  const checkIgnored: string = gitCheckIgnored(
    rushConfiguration.rushJsonFolder,
    finalTarballPath
  );
  if (checkIgnored) {
    throw new Error(`Tarball is ignored by git: ${checkIgnored}`);
  }
  spinner.succeed(`Tarball is not ignored by git`);

  // remove project config in rush.json
  spinner = ora(`Removing project config from rush.json`).start();
  rawRushJson.projects = rawRushJson.projects.filter(
    (x: JsonObject) => x.packageName !== packageName
  );
  JsonFile.save(rawRushJson, rushConfiguration.rushJsonFile, {
    updateExistingFile: true,
  });
  spinner.succeed(`Removed project config from rush.json`);

  // delete project folder
  spinner = ora(`Deleting project folder ${projectRelativeFolder}`).start();
  FileSystem.deleteFolder(projectFolder);
  spinner.succeed(`Deleted project folder ${projectRelativeFolder}`);
}
