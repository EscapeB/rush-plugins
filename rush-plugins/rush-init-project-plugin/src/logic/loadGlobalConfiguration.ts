import * as path from 'path';
import { LilconfigResult, lilconfigSync } from 'lilconfig';
import { FileSystem, Terminal, JsonFile, JsonSchema, JsonObject } from '@rushstack/node-core-library';
import templeteSchema from './template.schema.json';
import { TerminalSingleton } from '../terminal';
import { getTemplatesFolder, getTemplateNameList, ITemplatePathNameType } from './templateFolder';
import { IHooks } from '../hooks';
import { IPluginContext } from './TemplateConfiguration';

export interface IGlobalConfig {
  templateNameList: ITemplatePathNameType[];
}

const globalPluginLoaders: Record<string, (path: string) => void> = {
  '.js': (path) => require(path),
  '.ts': (path) => {
    require('ts-node').register({
      transpileOnly: true,
      compilerOptions: {
        esModuleInterop: true
      }
    });
    const module: any = require(path);
    try {
      if ('default' in module) {
        return module.default;
      }
    } catch {
      // no-catch
    }
    return module;
  }
};

// global configuration will load config from `template.json`
export async function loadGlobalConfiguration(
  hooks: IHooks,
  pluginContext: IPluginContext
): Promise<IGlobalConfig> {
  const terminal: Terminal = TerminalSingleton.getInstance();
  const templateFolder: string = getTemplatesFolder();
  const templateConfigPath: string = path.join(templateFolder, 'template.json');
  // if template.json doesn't exists, get the template list from folder
  if (!FileSystem.exists(templateConfigPath)) {
    return {
      templateNameList: getTemplateNameList(templateFolder)
    };
  }
  const jsonSchema: JsonSchema = JsonSchema.fromLoadedObject(templeteSchema);
  const res: JsonObject = await JsonFile.loadAndValidateWithCallbackAsync(
    templateConfigPath,
    jsonSchema,
    ({ details = '' }): void => {
      terminal.writeErrorLine(details);
      process.exit(1);
    }
  );
  const { templates, globalPluginEntryPoint } = res;
  if (globalPluginEntryPoint) {
    const result: LilconfigResult = lilconfigSync('init', {
      searchPlaces: [globalPluginEntryPoint],
      loaders: globalPluginLoaders
    }).search(templateFolder);
    if (result) {
      for (const plugin of result.config?.plugins) {
        plugin.apply(hooks, pluginContext);
      }
    }
  }

  return {
    templateNameList: templates as ITemplatePathNameType[]
  };
}
