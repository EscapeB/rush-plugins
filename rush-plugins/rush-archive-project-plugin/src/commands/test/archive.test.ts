import * as path from "path";
import { Executable } from "@rushstack/node-core-library";
import { archive } from "../archive";
import { defaultGraveyardRelativeFolder } from "../../logic/graveyard";

const folderSet = new Set<string>();

const resume = (folders: string[]) => {
  for (const folder of folders) {
    Executable.spawnSync("git", ["checkout", "."], {
      currentWorkingDirectory: folder,
    });
    Executable.spawnSync("git", ["clean", "-xdf"], {
      currentWorkingDirectory: folder,
    });
  }
};

describe("archive", () => {
  afterAll(() => {
    resume(Array.from(folderSet));
  });

  describe("archive basic", () => {
    const fixtureMonorepoPath = path.join(
      __dirname,
      "../../../",
      "fixtures/monorepo-1"
    );
    const fixtureMonorepoGraveyardPath = path.join(
      fixtureMonorepoPath,
      defaultGraveyardRelativeFolder
    );
    folderSet.add(fixtureMonorepoPath);
    folderSet.add(fixtureMonorepoGraveyardPath);
    beforeEach(() => {
      jest.spyOn(process, "cwd").mockReturnValue(fixtureMonorepoPath);
    });
    afterEach(() => {
      jest.resetAllMocks();
    });

    it("should not work with non existing package name", async () => {
      const packageName = "non-exist";
      await expect(
        archive({
          packageName,
          gitCheckpoint: false
        })
      ).rejects.toThrow(
        `Could not find project with package name ${packageName}`
      );
    });

    describe("fixture monorepo", () => {
      it("should work with fixture demo", async () => {
        const packageName = "demo-project";
        await expect(
          archive({
            packageName,
            gitCheckpoint: false
          })
        ).resolves.toBeUndefined();
      });
    });
  });

  describe("archive package with scope", () => {
    const fixtureMonorepoPath = path.join(
      __dirname,
      "../../../",
      "fixtures/monorepo-3"
    );
    const fixtureMonorepoGraveyardPath = path.join(
      fixtureMonorepoPath,
      defaultGraveyardRelativeFolder
    );
    folderSet.add(fixtureMonorepoPath);
    folderSet.add(fixtureMonorepoGraveyardPath);
    beforeEach(() => {
      jest.spyOn(process, "cwd").mockReturnValue(fixtureMonorepoPath);
    });
    afterEach(() => {
      jest.resetAllMocks();
    });

    it("should work with scoped package @my-company/my-app", async () => {
      const packageName = "@my-company/my-app";
      await expect(
        archive({
          packageName,
          gitCheckpoint: false
        })
      ).resolves.toBeUndefined();
    });
  });

  describe("archive package with dependent", () => {
    const fixtureMonorepoPath = path.join(
      __dirname,
      "../../../",
      "fixtures/monorepo-5"
    );
    const fixtureMonorepoGraveyardPath = path.join(
      fixtureMonorepoPath,
      defaultGraveyardRelativeFolder
    );
    folderSet.add(fixtureMonorepoPath);
    folderSet.add(fixtureMonorepoGraveyardPath);
    beforeEach(() => {
      jest.spyOn(process, "cwd").mockReturnValue(fixtureMonorepoPath);
    });
    afterEach(() => {
      jest.resetAllMocks();
    });

    it("should not work with package has dependent", async () => {
      const packageName = "@my-company/my-lib";
      await expect(
        archive({
          packageName,
          gitCheckpoint: false
        })
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
