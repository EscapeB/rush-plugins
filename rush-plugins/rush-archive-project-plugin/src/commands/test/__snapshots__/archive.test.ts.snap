// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`archive archive package with dependent should not work with package has dependent 1`] = `
"Target project @my-company/my-lib is depended by other 1 project(s):
@my-company/my-app"
`;
